include/sleip/dynamic_array.hpp

```diff
@@ -0,0 +1,320 @@
+#ifndef SLEIP_DYNAMIC_ARRAY_HPP_
+#define SLEIP_DYNAMIC_ARRAY_HPP_
+
+#include <boost/core/alloc_construct.hpp>
+#include <boost/throw_exception.hpp>
+#include <boost/config.hpp>
+
+#include <cstddef>
+#include <iterator>
+#include <memory>
+#include <utility>
+#include <type_traits>
+
+namespace sleip
+{
+template <class T, class Allocator>
+struct dynamic_array;
+
+template <class T, class Allocator = std::allocator<T>>
+struct dynamic_array
+{
+public:
+  using value_type             = T;
+  using allocator_type         = Allocator;
```

Comment:
Should check `typename Allocator::value_type` is `value_type`.

(see http://eel.is/c++draft/containers#tab:container.alloc.req)